cmd/vulcan-subdomain-takeover/main.go

```diff
@@ -0,0 +1,381 @@
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/route53"
+	"github.com/aws/aws-sdk-go-v2/service/route53/types"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/sirupsen/logrus"
+)
+
+const checkName = "vulcan-subdomain-takeover"
+
+// ErrTargetUnreachable means that it couldn't possible to reach the AWS target account.
+var ErrTargetUnreachable = errors.New("target is Unreachable")
+
+func main() {
+	run := func(ctx context.Context, target, assetType, optJSON string, state checkstate.State) error {
+		logger := check.NewCheckLogFromContext(ctx, checkName)
+
+		if target == "" {
+			return fmt.Errorf("check target missing")
+		}
+		scan, err := NewScanner(logger, target)
+		if err != nil {
+			return fmt.Errorf("could not create scan: %w", err)
+		}
+		takeovers, err := scan.Run()
+		if err != nil {
+			return fmt.Errorf("scan: %w", err)
+		}
+		addVulnsToState(state, takeovers, target)
+		return nil
+	}
+	c := check.NewCheckFromHandler(checkName, run)
+	c.RunAndServe()
+}
+
+type awsConfigurator interface {
+	getConfig(logger *logrus.Entry, target string) (aws.Config, error)
+}
+
+// route53Client represents a AWS route 53 client.
+type route53Client interface {
+	ListHostedZones(context.Context, *route53.ListHostedZonesInput, ...func(*route53.Options)) (*route53.ListHostedZonesOutput, error)
+	ListResourceRecordSets(ctx context.Context, params *route53.ListResourceRecordSetsInput, optFns ...func(*route53.Options)) (*route53.ListResourceRecordSetsOutput, error)
+}
+
+// ec2Client represents a AWS route EC2 client.
+type ec2Client interface {
+	DescribeAddressesAttribute(context.Context, *ec2.DescribeAddressesAttributeInput, ...func(*ec2.Options)) (*ec2.DescribeAddressesAttributeOutput, error)
+	DescribeNetworkInterfaces(ctx context.Context, params *ec2.DescribeNetworkInterfacesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeNetworkInterfacesOutput, error)
+}
+
+// ipRangesClient represents a IPRanges Client.
+type ipRangesClient interface {
+	GetPrefixes() (AWSPrefixes, error)
+}
+
+// Scanner represents a subdomain takeover scanner.
+type Scanner struct {
+	logger         *logrus.Entry
+	target         string
+	configurator   awsConfigurator
+	route53Client  route53Client
+	ec2Client      ec2Client
+	ipRangesClient ipRangesClient
+}
+
+// aswConfig represents the AWS configuration.
+type awsConfig struct{}
+
+// NewScanner creates a new instance of the Scanner.
+func NewScanner(logger *logrus.Entry, target string) (Scanner, error) {
+	cfg, err := awsConfig{}.getConfig(logger, target)
+	if err != nil {
+		return Scanner{}, fmt.Errorf("get config: %w", err)
+	}
+	return Scanner{
+		logger:         logger,
+		target:         target,
+		route53Client:  route53.NewFromConfig(cfg),
+		ec2Client:      ec2.NewFromConfig(cfg),
+		ipRangesClient: NewAWSIPRanges(),
+	}, nil
+}
+
+// Run executes the scan.
+func (s Scanner) Run() (map[string]string, error) {
```

Comment:
The process has an issue because the DNS records are global, but the getIPs is just for one region.
So it would report a False positive if one DNS record points to an IP on other region.